src/Functions/array/arrayIndex.h

```diff
@@ -574,94 +624,247 @@ struct ArrayIndexGenericNullImpl
     }
 };
 
-
-inline bool allowArrayIndex(const DataTypePtr & type0, const DataTypePtr & type1)
+/**
+ * Check types extracted from Nullable() and LowCardinality()
+ */
+inline bool allowArguments(const DataTypePtr & array_inner_type, const DataTypePtr & arg)
 {
-    DataTypePtr data_type0 = removeNullable(type0);
-    DataTypePtr data_type1 = removeNullable(type1);
-
-    return ((isNativeNumber(data_type0) || isEnum(data_type0)) && isNativeNumber(data_type1))
-        || data_type0->equals(*data_type1);
+    /**
+     * Possible cases for #arg and #array_inner_type:
+     * 1. T
+     * 2. LC(T)
+     * 3. N(T)
+     * 4. N(LC(T)) -- quite strange, rarely found but possible.
+     * 5. LC(N(T))
+     *
+     * All other variants are considered wrong (Like N(N(N(T)))) or LC(N(LC(T))).
+     * recursiveRemoveLowCardinality works only if the given type is LC(V).
+     */
+    DataTypePtr array_extracted =
+        removeNullable(                    /// remove outer Nullable, cases 3 and 4
+            recursiveRemoveLowCardinality( /// remove LC, cases 2 and 4
+                removeNullable(            /// remove inner Nullable, cases 3 and 5
+                    array_inner_type)));
+
+    DataTypePtr arg_extracted =
+        removeNullable(
+            recursiveRemoveLowCardinality(
+                removeNullable(
+                    arg)));
+
+    return ((isNativeNumber(array_extracted) || isEnum(array_extracted)) && isNativeNumber(arg_extracted))
+        || array_extracted->equals(*arg_extracted);
 }
 
-
-template <typename IndexConv, typename Name>
+template <class ConcreteAction, class Name>
 class FunctionArrayIndex : public IFunction
 {
 public:
     static constexpr auto name = Name::name;
     static FunctionPtr create(const Context &) { return std::make_shared<FunctionArrayIndex>(); }
 
 private:
-    using ResultColumnType = ColumnVector<typename IndexConv::ResultType>;
+    using ResultType = typename ConcreteAction::ResultType;
+    using ResultColumnType = ColumnVector<ResultType>;
+
+    /**
+     * The Array's internal data type may be quite tricky (containing a Nullable type somewhere). To process the 
+     * Nullable types correctly, for each data type specialisation we provide two null maps (one for the data and one 
+     * for the items). By convention they are passed as the third and the fourth argument, respectively 
+     * (counting from 1).
+     *
+     * @return {nullptr, nullptr} if there are less then 3 arguments.
+     * @return {null_map_data, nullptr} if there are three arguments
+     * @return {nullptr, null_map_item} if there are four arguments but the third is missing.
+     * @return {null_map_data, null_map_item} if there are four arguments.
+     */
+    std::pair<
+        const PaddedPODArray<UInt8> *,
+        const PaddedPODArray<UInt8> *>
+        nullMapsBuilder(const Block& block, const ColumnNumbers & arguments) const noexcept
```

Comment:
Why is it 'builder' if it doesn't build anything? at most, it 'gets' or 'extracts''